File: src/main/java/com/pedigree/editor/commands/AttachMediaToFamilyCommand.java
package com.pedigree.editor.commands;

import com.pedigree.editor.Command;
import com.pedigree.model.Family;
import com.pedigree.model.MediaAttachment;
import com.pedigree.storage.ProjectRepository;

import java.util.Objects;

public class AttachMediaToFamilyCommand implements Command {
    private final ProjectRepository.ProjectData data;
    private final String familyId;
    private final MediaAttachment media;
    private boolean attached;

    public AttachMediaToFamilyCommand(ProjectRepository.ProjectData data, String familyId, MediaAttachment media) {
        this.data = Objects.requireNonNull(data, "data");
        this.familyId = Objects.requireNonNull(familyId, "familyId");
        this.media = Objects.requireNonNull(media, "media");
    }

    @Override
    public void execute() {
        Family f = data.families.stream().filter(fam -> fam.getId().equals(familyId)).findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No family: " + familyId));
        boolean exists = f.getMedia().stream().anyMatch(m -> m.getId().equals(media.getId()));
        if (!exists) {
            f.getMedia().add(media);
            attached = true;
        }
    }

    @Override
    public void undo() {
        if (!attached) { return; }
        data.families.stream().filter(fam -> fam.getId().equals(familyId)).findFirst()
                .ifPresent(f -> f.getMedia().removeIf(m -> m.getId().equals(media.getId())));
        attached = false;
    }
}

File: src/main/kotlin/com/family/tree/editor/commands/AttachMediaToFamilyCommand.kt
package com.family.tree.editor.commands

import com.family.tree.editor.Command
import com.family.tree.storage.ProjectRepository
import java.util.Objects

class AttachMediaToFamilyCommand(
    private val data: ProjectRepository.ProjectData,
    private val familyId: String,
    private val media: com.family.tree.model.MediaAttachment
) : Command {
    private var attached = false

    init {
        Objects.requireNonNull(data, "data")
        Objects.requireNonNull(familyId, "familyId")
        Objects.requireNonNull(media, "media")
    }

    override fun execute() {
        val f = data.families.firstOrNull { it.id == familyId }
            ?: throw IllegalArgumentException("No family: $familyId")
        val exists = f.media.any { it.id == media.id }
        if (!exists) {
            f.media.add(media)
            attached = true
        }
    }

    override fun undo() {
        if (!attached) return
        data.families.firstOrNull { it.id == familyId }?.media?.removeIf { it.id == media.id }
        attached = false
    }
}
